### ISSUE TITLE
build: Detect and handle RAM errors

### ISSUE THREAD

**Issue by Lestropie** [MRtrix3/mrtrix3#660]
Spawned from #655.

If a compilation task fails due to running out of RAM, add it to a list of tasks that should be run single-threaded. Once all multi-threaded tasks that can be run have been run, execute the single-threaded tasks sequentially. Then, go back to multi-threading. Repeat until there are no tasks left.

Tagged as wishlist, but it seems more and more people are running into memory issues during compilation.

**Comment by jdtournier** [comment]
Yes, I've been thinking along similar lines. The problem as far as I can tell is detecting reliably when the task fell over because of RAM limitations rather than a genuine compiler error...

**Comment by draffelt** [comment]
We could just throttle the number of CPU cores based on the amount of
system RAM (and maybe compiler).

**Comment by jdtournier** [comment]
> We could just throttle the number of CPU cores based on the amount of
system RAM (and maybe compiler).

Sure, but it's going to be pretty fiddly to get right. Besides, I have a sneaky suspicion that planning for the worst-case may preclude using all cores generally, just on the off-chance that you end up compiling `tckgen`, `mrregister`, `mrconvert`, `mrmath`, and probably a few others all at the same time in debug mode. I prefer Rob's approach - but it does depend on the ability to reliably detect out of RAM failures...

**Comment by maxpietsch** [comment]
The simplest solution would be to have a second task queue for failed commands and repeat those single-threaded after compiling the rest. If they still fail then it's a compilation error or still not enough RAM. I can see that this is annoying in development but we could deactivate this with an option.

**Comment by draffelt** [comment]
Max, that could probably work, and would probably be sufficient for the few
machines that run out of RAM.

**Comment by jdtournier** [comment]
@maxpietsch: that wont be sufficient. We can't simply re-run those that failed, since there will undoubtedly be further jobs to run that depends on these. Most of the failures are going to be compilation of command cpp files, they'll need to be linked after that. This is why @Lestropie is suggesting iterating through again after that.

**Comment by maxpietsch** [comment]
@jdtournier I think this is very similar to what @Lestropie suggests but with commands I meant .cpp files in cmd/ not all build tasks. So allow compile jobs to fail and if linking of a command does not work, add it (and its dependent objects) to the single threaded queue.
This has the advantage that on low-memory systems ./build compiles all commands that are possible to compile with the given RAM and then retries in the most memory-conservative manner. If that still fails then one has at least the lighter commands. The disadvantage is that if a shared object fails to compile then all commands that use it will be added to the single threaded queue.

**Comment by jdtournier** [comment]
OK, I see the subtle distinction... So you'd basically rely on the assumption that only cpp files in the `cmd` folder can fail due to RAM constraints, and furthermore that when these compile jobs fail, it must be due to RAM limitations (at least on a non-dev system)? So this would avoid the need to detect failures due to RAM constraints...?

Another idea that's probably much simpler and can be implemented pretty much straight away, is to change the default behaviour so that `build` continues running through its list until all the jobs it can do have been at least attempted. If jobs did fail, it could then issue a warning suggesting to users to try running it again. Hopefully, the only jobs left to run again will be those causing trouble. They might compete with each other for the first round or two, but should eventually all compile OK. If jobs failed but no job was successful, then that means there's no point trying again, so just stick to the usual error in this case. Given that these issues hopefully happen relatively rarely, it's probably a good enough solution for now, and it certainly hints to the user that this is not normal behaviour, and maybe they should think about increasing their RAM (otherwise they probably won't be able to process HCP data, etc, anyway). Thoughts?
